File: packages/client/src/namespaces/bench.ts
//! Benchmark helpers (non-§12.1). The `Channel`-based stream benchmark lives
//! here so the transport import is isolated to one module.

import { Channel } from "@tauri-apps/api/core";
import type {
  BenchmarkConfig,
  BenchmarkResult,
  DiffHunk,
  StreamChunk,
} from "@tethys/bindings";

import type { Call } from "../transport";

export function benchFlatHelpers(call: Call) {
  return {
    /** `git.diff.synthetic` — S0.1 diff benchmark generator */
    getSyntheticDiff: (lineCount = 20000) =>
      call<DiffHunk[]>("generate_synthetic_diff", { lineCount }),
    /** `bench.stream` — S0.1 IPC channel streaming */
    runStreamBenchmark: async (
      config: BenchmarkConfig,
      onChunk: (chunk: StreamChunk) => void,
    ): Promise<BenchmarkResult> => {
      const channel = new Channel<StreamChunk>();
      channel.onmessage = onChunk;
      return call<BenchmarkResult>("run_stream_benchmark", {
        config,
        onChunk: channel,
      });
    },
  };
}

File: apps/desktop/src-tauri/src/commands/bench.rs
//! Benchmark helper commands (non-§12.1).

use tauri::ipc::Channel;
use tauri::State;
use tethys_api::BenchApi;
use tethys_schema::{BenchmarkConfig, BenchmarkResult, DiffHunk, StreamChunk};

use crate::commands::CoreState;

/// `git.diff.synthetic` — S0.1 diff benchmark generator.
#[tauri::command]
#[specta::specta]
pub async fn generate_synthetic_diff(
    state: State<'_, CoreState>,
    line_count: usize,
) -> Result<Vec<DiffHunk>, String> {
    state
        .generate_synthetic_diff(line_count)
        .await
        .map_err(|e| e.to_string())
}

/// `bench.stream` — S0.1 IPC channel streaming.
#[tauri::command]
pub async fn run_stream_benchmark(
    config: BenchmarkConfig,
    on_chunk: Channel<StreamChunk>,
) -> Result<BenchmarkResult, String> {
    let start = std::time::Instant::now();
    let total_per_stream = config.total_messages / config.streams.max(1);
    let mut handles = Vec::new();

    for stream_id in 0..config.streams {
        let channel = on_chunk.clone();
        let payload_size = config.message_size_bytes as usize;
        let chunks = tethys_core::synthetic::generate_stream_chunks(
            stream_id,
            total_per_stream,
            payload_size,
        );

        handles.push(tokio::spawn(async move {
            for chunk in chunks {
                let _ = channel.send(chunk);
            }
        }));
    }

    for h in handles {
        let _ = h.await;
    }

    let elapsed = start.elapsed();
    let elapsed_ms = elapsed.as_secs_f64() * 1000.0;
    let msgs_per_sec = (config.total_messages as f64) / elapsed.as_secs_f64().max(0.001);

    Ok(BenchmarkResult {
        total_messages: config.total_messages,
        elapsed_ms,
        messages_per_sec: msgs_per_sec,
        p95_latency_ms: elapsed_ms / (config.total_messages as f64).max(1.0),
    })
}
